crates/astria-conductor/src/conductor/inner.rs

```diff
@@ -162,51 +154,56 @@ impl Conductor {
         };
 
         Ok(Self {
-            shutdown,
+            shutdown_token,
             tasks,
         })
     }
 
-    /// Runs [`Conductor`] until it receives an exit signal.
+    /// Runs [`ConductorInner`] until it receives an exit signal.
     ///
     /// # Panics
     /// Panics if it could not install a signal handler.
-    async fn run_until_stopped(mut self) {
+    async fn run_until_stopped(mut self) -> eyre::Result<RestartOrShutdown> {
         info_span!("Conductor::run_until_stopped").in_scope(|| info!("conductor is running"));
 
         let exit_reason = select! {
```

Comment:
For my suggestion further down: you don't need to keep the type of this `exit_reason` the same.

You can just introduce an enum like this:

```rust
enum ExitReason {
    ShutdownSignal,
    TaskFailed { name: &'static str, error: eyre::Report, }
}
```

And then your `shutdown` logic can deal with this elegantly.